[WIP] Example of multiple imputation with IterativeImputer

sergeyf opened the pull request:
Adding to #11977. This PR is a restart of #11370, which got messy.

Here is a quote from #11370 that explains what this PR does:

> This PR is an example that shows how to use IterativeImputer for Multiple Imputation.
> 
> As discussed in #11259, the defaults of IterativeImputer are such that single imputation is performed. Because the method is also quite powerful for Multiple Imputation, we agreed to make an example that shows the user how to use ImputerImputer to perform Multiple Imputation.
> 
> I made the document: examples/impute/plot_multiple_imputation.py and it shows 2 things:
> 
> Estimation of beta estimates and their standard error: compare IterativeImputer with using IterativeImputer as a MICE Imputer.
> How to use IterativeImputer as a MICE Imputer when making a prediction model (with train and test datasets).

jnothman committed:
FEA Reinstate ChainedImputer

This reverts commit f819704880b5d6affb49996f832f1aa5c8799571.

jnothman committed:
Fix import of time

jnothman committed:
Merge branch 'master' into iterativeimputer

sergeyf committed:
[MRG] ChainedImputer -> IterativeImputer, and documentation update (#11350)

Towards making this more generic than MICE

benlawson committed:
[MRG] sample from a truncated normal instead of clipping samples from a normal (#12177)

jnothman committed:
Merge branch 'master' into iterativeimputer

jnothman committed:
DOC Merge IterativeImputer what's news

jnothman committed:
Merge branch 'master' into iterativeimputer

jnothman committed:
Undo changes to v0.20.rst

jnothman committed:
Revert changes to v0.20.rst

jnothman committed:
DOC Normalize whitespace in doctest

jnothman committed:
Fix for SciPy 0.17

jnothman committed:
Fix doctest

Commit:
first commit, author=Rianne

sergeyf commented:
Paing @jnothman and @RianneSchouten.

sergeyf committed:
flakes

sergeyf committed:
adding authors

jnothman commented:
It might be good to amend that first commit with `--author 'Rianne Schouten'` etc. Thanks for this! Will look soon!

sergeyf commented:
OK, I think that worked.

jnothman commented:
I suspect that without  `examples/impute/README.txt` the example won't render.

jnothman committed:
Create examples/impute gallery

jnothman committed:
Merge branch 'iterativeimputer' into HEAD

sergeyf committed:
maybe this will fix the 24/26 issue?

sergeyf committed:
adding readme

jnothman commented:
Oh, no. Did I break the doctest again?

jnothman committed:
Add missing readme file

jnothman committed:
Undo change to circle build

sergeyf committed:
trying to fix the weird impute.rst problem

sergeyf committed:
Merge branch 'iterativeimputer' into iterativeimputer_mice_example

sergeyf commented:
I don't know what's up with that thing. I'm trying to modify the test to see if a different test in the `.predict` call will make it stop complaining.

sergeyf committed:
fewer iters to avoid timeout

sergeyf committed:
Merge branch 'iterativeimputer_mice_example' of https://github.com/sergeyf/scikit-learn into iterativeimputer_mice_example

jnothman commented:
2 minutes is still quite long for the example gallery

>
>

sergeyf commented:
Hmm, it's the nature of the example. MICE just runs stuff a lot of times. I can keep reducing the numbers of iterations, etc. Should we get rid of one of the two parts of this example?

sergeyf committed:
fewer iters to speed up example

sergeyf commented:
Ok that didn't work. To be clear, it looks like it is actually finishing:

```
generating gallery for auto_examples/impute... [100%] plot_multiple_imputation.py
Too long with no output (exceeded 10m0s)
```

So it goes to 100% and then times out?

jnothman commented:
Yeah, it runs quite quickly on my machine too. Not sure yet what's going
on, but I've not yet reviewed the code

jnothman committed:
DOC Make IterativeImputer doctest more stable (#13026)

jnothman committed:
TST IterativeImputer: Check predictor type (#13039)

sergeyf committed:
Merge branch 'iterativeimputer' into iterativeimputer_mice_example

sergeyf commented:
Alright, I'll pause here until you've had a chance to glance at it. The failure doesn't make sense to me.

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+
+
+# Start by defining a basic amputation function
+def ampute(X, missing_rate=0.75, mech="MCAR"):
```
Please add a brief docstring at least describing what ampute means and MCAR vs MNAR

jnothman replied:
rewrite this function:
```suggestion
def ampute(X, missing_rate=0.75, mech="MCAR"):
	X_out = X.copy()
	n_drop_per_feature = int(missing_rate * X.shape[0])
	for x in np.transpose(X_out):
	    # insert missing values for each feature
	    if mech == 'MCAR':
	        prob = None
	    else:
	        weights = scipy.special.expit(x - x.mean())
	        prob = weights / weights.sum()
	    drop_idx = np.random.choice(X.shape[0], p=prob, replace=False,
	                                size=n_drop_per_feature)
	    x[drop_idx] = np.nan
	return X_out
```

sergeyf replied:
We don't even use MNCAR in this example because it's so computationally expensive. I'll simplify and remove references to MNCAR.

sergeyf replied:
Well, I guess it might be useful to have if it's only a few lines of code as you rewrote it.

glemaitre replied:
@jnothman Is there a plan to have an ``ampute`` function in ``sklearn.datasets``?

glemaitre replied:
I would think about `strategy` instead of `mech`?

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+    if mech == "MNAR":
+        for i in np.arange(n_features):
+            data_values = -np.mean(X[:, i]) + X[:, i]
+            weights = 1 / (1 + np.exp(-data_values))
```
Use scipy.special.expit

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+        for i in np.arange(n_features):
+            data_values = -np.mean(X[:, i]) + X[:, i]
+            weights = 1 / (1 + np.exp(-data_values))
+            probs = np.array(weights) / np.sum(np.array(weights))
```
Don't need these np.array

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+    # MNAR mechanism
+    if mech == "MNAR":
+        for i in np.arange(n_features):
+            data_values = -np.mean(X[:, i]) + X[:, i]
```
Call this x_centred.

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+
+
+# Make a function that calculates the variance of the beta estimates. This is
+# necessary because the linear regression model from sklearn does not provide
```
I think we can assume without stating it that everything in the example is not in the library.

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+# these values.
+def calculate_variance_of_beta_estimates(y_true, y_pred, X):
+    residuals = np.sum((y_true - y_pred)**2)
+    sigma_hat_squared = (1 / (len(y_true) - 2)) * residuals
```
```suggestion
    sigma_hat_squared = residuals / (len(y_true) - 2)
```

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+# necessary because the linear regression model from sklearn does not provide
+# these values.
+def calculate_variance_of_beta_estimates(y_true, y_pred, X):
+    residuals = np.sum((y_true - y_pred)**2)
```
This is a sum of squared errors... should not be plural "residuals". Better name?

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+    # confidence interval.
+    full_coefs = estimator.coef_
+    full_vars = calculate_variance_of_beta_estimates(y, y_predict, X)
+    full_errorbar = 1.96 * np.sqrt(full_vars)
```
Calculate this in plotting rather than all the results functions

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+    # Perform linear regression on chained single imputed data
+    # Estimate beta estimates and their variances
+    estimator = RidgeCV()
+    estimator.fit(X_imputed, y)
```
use fit_transform

sergeyf replied:
RidgeCV doesn't have fit_transform?

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+    m = 5
+    multiple_imputations = []
+    for i in range(m):
+        imputer = IterativeImputer(n_iter=N_ITER,
```
Just call `get_results_chained_imputation(X_incomplete, y, random_state=i)` then do the pooling

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+# estimates (B).
+#
+# Make a function that calculates Qbar from m estimates
+def calculate_Qbar(m_estimates):
```
rename to `pool_coefs`

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+
+
+# Make a function that calculates T from m estimates and their variances
+def calculate_T(m_estimates, m_variances, Qbar):
```
rename to `pool_variances`

jnothman reviewed (commented):
This still needs some work :)

jnothman committed:
Update examples/impute/plot_multiple_imputation.py

Joel's Ampute modification.

Co-Authored-By: sergeyf <sergeyfeldman@gmail.com>

sergeyf committed:
addressing review comments

sergeyf committed:
flake

sergeyf commented:
@jnothman I've addressed all of your comments, except the fit_transform one that doesn't quite make sense to me. After reading it over, any idea why it times out on the test?

sergeyf committed:
EHN: Changing default model for IterativeImputer to BayesianRidge (#13038)

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+rng = np.random.RandomState(0)
+
+N_ITER = 20  # number of iterations in IterativeImputer
+N_SIM = 3  # number of simulations in Example 2
```
For an example, I think that we could be a bit more flexible with PEP8. Move this variable closer to its use even if defined globally.

sergeyf replied:
I put it up here so it's easy to tell what you can change to make it run much faster. I'd prefer to leave as is.

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+def ampute(X, missing_rate=0.75, mech="MCAR"):
+    """Ampute (opposite of impute) function to simulate missing data in X
+
+    MCAR - missing completely at random
```
Let's put a real docstring

```python
"""Insert missing data in X.

Ampute is the inverse of impute and serve at simulating a dataset with missing data. Several strategies exist to remove data: missing completely at random (MCAR) and missing not completely at random (MNCAR).

Parameters
----------
X : ndarray, shape (n_samples, n_features)
    The data without missing samples.
missing_rate : float, default=0.75
    The amount of missing data in ``X``.
mech : str, default='MCAR'
    The strategy for removing data.

Returns
-------
X_out : ndarray, shape (n_samples, n_features)
    Output data with missing entries.
```

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+Imputing missing values using multiple imputation
+=================================================
+
+By default, the `sklearn.impute.IterativeImputer` performs single imputation:
```
` :class:`sklearn.impute.IterativeImputer` `

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+        if mech == 'MCAR':
+            prob = None
+        else:
+            weights = special.expit(x - x.mean())
```
It might be interesting to mention what is going here. Maybe either a commend or in the docstring of the function. From what I understand, we put some weight such that we have more chance to remove values close to the mean?

sergeyf replied:
Well not quite since expit is not symmetric. So if you're much smaller than the mean you are unlikely to be removed, but if you're much larger than the mean than you are very likely to be removed. This is confusing, so I'll change it to np.abs(x - x.mean()) and then it'll be as you explained,

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+    sigma_hat_squared = sum_sq_errors / (len(y_true) - 2)
+    X_prime_X = np.dot(X.T, X)
+    covariance_matrix = sigma_hat_squared / X_prime_X
+    vars = np.diag(covariance_matrix)
```
```suggestion
    return np.diag(covariance_matrix)
```

It avoids to redefine the vars function which is a built-in function

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+    return X_out
+
+
+# Make a function that calculates the variance of the beta estimates.
```
Make a proper docstring instead of a comment above the function

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+def calculate_variance_of_beta_estimates(y_true, y_pred, X):
+    sum_sq_errors = np.sum((y_true - y_pred)**2)
+    sigma_hat_squared = sum_sq_errors / (len(y_true) - 2)
+    X_prime_X = np.dot(X.T, X)
```
remove this line

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+    sum_sq_errors = np.sum((y_true - y_pred)**2)
+    sigma_hat_squared = sum_sq_errors / (len(y_true) - 2)
+    X_prime_X = np.dot(X.T, X)
+    covariance_matrix = sigma_hat_squared / X_prime_X
```
```suggestion
    covariance_matrix = sigma_hat_squared / np.dot(X.T, X)
```

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+# Make a function that calculates the variance of the beta estimates.
+def calculate_variance_of_beta_estimates(y_true, y_pred, X):
+    sum_sq_errors = np.sum((y_true - y_pred)**2)
+    sigma_hat_squared = sum_sq_errors / (len(y_true) - 2)
```
By curiosity, why do we need `- 2`?

sergeyf replied:
https://stats.stackexchange.com/questions/204238/why-divide-rss-by-n-2-to-get-rse

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+#
+# Make a function that calculates Qbar from m estimates
+def pool_coefs(m_estimates):
+    m = len(m_estimates)
```
```suggestion

```

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+def get_results_mice_imputation(X_incomplete, y, m=5):
+    # Perform a model on each of the m imputed datasets
+    # Estimate the estimates for each model/dataset
+    m_coefs = []
```
You should be able to make a comprehension here
```python
m_coefs, m_vars = zip(*[get_results_chained_imputation(...) for i in range(m)])
```

sergeyf replied:
Eh it's simple enough. I don't think this will save any compute.

glemaitre replied:
I disagree. List comprehensions are pythonic and even if this is difficult to understand at first, this is a recurrent pattern used by Python programmers a lot. I think that our example needs to use good programming practice. However, if it requires a comment explaining what it does, I am not opposed to that.

However, this is a personal opinion, so I don't want to force it on the review. I'll let another reviewer decide which syntax would be the best for the example.

sergeyf replied:
Ok, I'm convinced =)

chkoar replied:
I agree with @glemaitre regarding the list comprehensions

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+    return chained_coefs, chained_vars
+
+
+def get_results_mice_imputation(X_incomplete, y, m=5):
```
`m` is not enough explicit name

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+# Make a function that calculates Qbar from m estimates
+def pool_coefs(m_estimates):
+    m = len(m_estimates)
+    Qbar = np.sum(m_estimates, axis=0) / m
```
```suggestion
    Qbar = np.sum(m_estimates, axis=0) / len(m_estiamtes)
```

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+# estimates (B).
+#
+# Make a function that calculates Qbar from m estimates
+def pool_coefs(m_estimates):
```
We need to have in the docstring that we play with  python list.

sergeyf replied:
Sorry can you clarify what this means?

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,392 @@
+
+# Make a function that calculates T from m estimates and their variances
+def pool_variances(m_estimates, m_variances, Qbar):
+    m = len(m_estimates)
```
m is not really explicit

glemaitre reviewed (changes requested):
I'll continue my review later.

sergeyf committed:
partial reviewer comments

sergeyf committed:
Merge branch 'iterativeimputer' into iterativeimputer_mice_example

sergeyf committed:
simplifying code a bit

sergeyf committed:
more docs

sergeyf committed:
better var name

glemaitre commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,396 @@
+# every i in m and then calculate the evaluation metric. We test both
+# approaches.
+
+# Apply the regression model on the full dataset as a way of comparison.
```
I think that we need to reduce the number of functions. I don't think that we actually need any function if we use the pipeline and set the random state whenever this is necessary.

sergeyf replied:
Sorry, how broadly does this apply? What does "any function" refer to?

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,396 @@
+    imputer = IterativeImputer(n_iter=N_ITER,
+                               sample_posterior=True,
+                               random_state=random_state)
+    imputer.fit(X_incomplete)
```
This is the one I meant

jnothman reviewed (commented):
Forgot to submit these

sergeyf commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,396 @@
+
+
+def rubins_pooling_rules(m_estimates, m_variances):
+    """Applies Rubin's pooling rules.
```
It matches all the docstring explanation and makes it easier to follow, I think.

sergeyf commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,396 @@
+X_full, y = dataset.data, dataset.target
+
+# Standardizing the data
+scaler = StandardScaler()
```
I'll put this on top and it can be reused everywhere
```ESTIMATOR = Pipeline(steps=[('scaler', StandardScaler),
                            ('bayesianridge', BayesianRidge)])```

sergeyf commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,396 @@
+# Standardizing the data
+scaler = StandardScaler()
+X_scaled = scaler.fit_transform(X_full)
+y_scaled = stats.zscore(y)
```
I think it's mostly to affect the scale of the coefs. Not a big deal really.

sergeyf commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+plt1 = plt.subplot(211)
+for j in n:
+    plt1.bar(np.arange(len(coefs[j])) + (3*j*(width/n_situations)),
+             coefs[j], width=width, color=colors[j])
```
I think there must be some confusion due to the variable naming. The two figures that are part of example 1 do not have error bars because they are the result of a single run. `errorbars` are actually the standard errors.

sergeyf commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,457 @@
+# situation. There are two approaches to get the end result of the prediction
+# model. In approach 1 you calculate the evaluation metric for every i in m and
+# later average these values. In approach 2 you average the predictions of
+# every i in m and then calculate the evaluation metric. We test both
```
It's true, we just do approach 2. To my eyes, the ML audience will care more about the simulation than the standard error analysis. 

How do you propose we look at the variability in prediction and MSE due to MI?

sergeyf commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,396 @@
+# variable. The reason to do this is that the imputation model should at least
+# contain the analysis model to result in unbiased estimates. In this function,
+# we will also include y in the imputation process.
+def get_results_mice_imputation_includingy(X_incomplete, y, num_imputations=5):
```
Actually we do because the y should only participate in the imputation, but not in the subsequent linear regression.

sergeyf replied:
I refactored a bit.

sergeyf committed:
lots of review fixes

jnothman reviewed (commented):
I don't, btw think there is anything wrong with defining functions, as long as choose duplication is minimised and the control flow is clearly readable

sergeyf committed:
jnothman's better docstring

jnothman commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -3,23 +3,19 @@
+:class:`sklearn.impute.IterativeImputer` uses a different `random_state`, this
+results in multiple imputations, each of which can be used to train a
+predictive model.
 
```
I'm not sure how much of the rest we need now either

sergeyf replied:
Well, the Rubin's stuff is pretty new to me, so I think it's worthwhile to keep it in.

sergeyf committed:
EXA Add IterativeImputer extended example (#12100)

glemaitre commented:
> I don't, btw think there is anything wrong with defining functions, as long as choose duplication is minimised and the control flow is clearly readable

I actually don't mind the duplication. Here my argument is that if the function is actually creating a pipeline, fitting, and getting the score, I don't think that introducing an indirection in the reading of the example is necessary. I prefer to create comment blocks around those steps which will create a cell in the notebook rendering online.

glemaitre reviewed (changes requested):
Partial review.

sergeyf committed:
Merge branch 'iterativeimputer' into iterativeimputer_mice_example

sergeyf commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,367 @@
+                                    ('regressor', BayesianRidge())])
+
+
+def ampute(X, missing_rate=0.75, mech="MCAR"):
```
Thanks for doing that. I'll stick it in.

sergeyf commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,367 @@
+    return np.diag(covariance_matrix)
+
+
+def rubins_pooling_rules(m_estimates, m_variances):
```
Thanks, I'll take it.

sergeyf commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,367 @@
+                               sample_posterior=True,
+                               random_state=random_state)
+    if impute_with_y:
+        Xy = np.column_stack((X_incomplete, y_scaled))
```
As you suggest, I'll get rid of y_scaled everywhere.

sergeyf committed:
more reviewer comments

sergeyf commented:
@glemaitre Thanks, especially for the fully written functions. I have addressed many more comments.

sergeyf committed:
correcting comment

sergeyf committed:
trailing white space =(

jnothman commented:
This remains tricky to push to an ML audience, where we measure the uncertainty over the whole pipeline through cross-validation. I still think it needs some work to speak to that audience, for instance not referring to "estimates" without clarifying that these are coefficient/parameter estimates. We've improved it a lot, and I'll give it more review, but I don't think it's critical to merging iterativeimputer into master.

sergeyf commented:
Ok, I'll switch to making the early stopping PR.

On Sat, Jan 26, 2019, 3:20 AM Joel Nothman <notifications@github.com wrote:

> This remains tricky to push to an ML audience, where we measure the
> uncertainty over the whole pipeline through cross-validation. I still think
> it needs some work to speak to that audience, for instance not referring to
> "estimates" without clarifying that these are coefficient/parameter
> estimates. We've improved it a lot, and I'll give it more review, but I
> don't think it's critical to merging iterativeimputer into master.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/13025#issuecomment-457823256>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABya7DFrfDVKf8xrOs5S0-2RjlTxLaZ7ks5vHDoTgaJpZM4aLYvi>
> .
>

sergeyf committed:
master merge

sergeyf committed:
undoing some changes left over from merge

sergeyf commented:
I've rebased this. If it fails the tests again, maybe someone can help me figure out why it's timing out on circleci but not on any local machines? I've made it so this example is relatively quick (~1m).

sergeyf committed:
n_iter fix

banilo commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,359 @@
+    """Insert missing data into X.
+
+    Ampute is the inverse of impute and serves at simulating a dataset with
+    missing data. Two strategies are implemented to remove data: (1) missing
```
not sure 'strategy' is the right word here. How about 'missing data mechanism' ?

banilo commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,359 @@
+# IMPUTATION IN A LINEAR REGRESSION MODEL.
+
+MAX_ITER = 20  # number of iterations in IterativeImputer
+ESTIMATOR = Pipeline(steps=[('scaler', StandardScaler()),
```
all capital case variable names for a class instance appears less common for the sklearn ecosystem...

sergeyf replied:
I'm doing this because it's a global for the entire file, but I'm open to other styling.

banilo commented on examples/impute/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,359 @@
+N_SIM = 3  # number of simulations in Example 2
+ESTIMATOR_IMPUTER = Pipeline(steps=[('imputer', IterativeImputer()),
+                                    ('scaler', StandardScaler()),
+                                    ('regressor', BayesianRidge())])
```
same comment as above; perhaps 'regression_model' ?

amueller commented:
What's the status of this PR and how does it relate to the convergence issues #14338?

sergeyf commented:
No relation to #14338 

Status: waiting for someone who actually knows the ins and outs of MICE to help here. Also, some tests were failing on the test machine but not locally and we failed to debug them.

sergeyf committed:
Merge branch 'master' into iterativeimputer_mice_example

sergeyf commented:
Ok @amueller the problem is that the failing test times out, but it takes way less time on a local machine than 10m. I have no idea why.

jnothman commented:
Perhaps we can rerun the job with ssh and go in to investigate...

sergeyf commented:
Is that something I have access to?

On Thu, Aug 8, 2019, 1:13 AM Joel Nothman <notifications@github.com> wrote:

> Perhaps we can rerun the job with ssh and go in to investigate...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/13025?email_source=notifications&email_token=AAOJV3ATMQGIHMEYLFVBIPLQDPISBA5CNFSM4GRNRPRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD322T3I#issuecomment-519416301>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOJV3D2XW6QKC3FNNVHNJTQDPISBANCNFSM4GRNRPRA>
> .
>

jnothman commented:
Perhaps if you enabled circle on your fork?

sergeyf commented:
I currently don't have the bandwidth for that brand of mucking about. Any
chance someone with access could take a look?

On Thu, Aug 8, 2019 at 3:14 PM Joel Nothman <notifications@github.com>
wrote:

> Perhaps if you enabled circle on your fork?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/13025?email_source=notifications&email_token=AAOJV3EK6VBEQVI7WFVQRPDQDSLF5A5CNFSM4GRNRPRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD35BPWA#issuecomment-519706584>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOJV3DK2VOEAJFIXDKXTNDQDSLF5ANCNFSM4GRNRPRA>
> .
>

sergeyf committed:
merge master

sergeyf committed:
Merge branch 'iterativeimputer_mice_example' of https://github.com/sergeyf/scikit-learn into iterativeimputer_mice_example

sergeyf committed:
Merge branch 'master' into iterativeimputer_mice_example

sergeyf committed:
undoing weird merge changes

sergeyf committed:
Merge branch 'master' into iterativeimputer_mice_example

sergeyf commented:
@jnothman I wondered back to this PR, and now it passes tests!

Any interest in picking up work on this? I feel like it was in a pretty good place already and we were just unsure about the extremely long runtimes.

sergeyf commented:
@jnothman @glemaitre Any thoughts on my last comment? Repeated here: "Any interest in picking up work on this? I feel like it was in a pretty good place already and we were just unsure about the extremely long runtimes."

nxorable commented:
I [found the current docs ambiguous](https://github.com/scikit-learn/scikit-learn/discussions/22103) and believe the community would value this work.

sergeyf commented:
I'm not familiar with these build trigger checks. Can anyone please suggest how to fix it?

thomasjpfan commented:
> Can anyone please suggest how to fix it?

Syncing with the `main` branch should fix the build trigger error.

sergeyf committed:
Merge branch 'scikit-learn:main' into iterativeimputer_mice_example

sergeyf commented:
Thank you @thomasjpfan! Any idea if we can get this merged once all the tests pass?

sergeyf committed:
run black

sergeyf committed:
black try again

sergeyf committed:
final black try. somehow "black file" did not do it all...

thomasjpfan commented:
> Any idea if we can get this merged once all the tests pass?

This PR still need two approvals to get merged and I do not have a time estimate for that to happen.

At a glance, I see two big tasks for this PR:

1. It needs to use another dataset instead of `load_boston` because `load_boston` has been deprecated and will be removed in 1.2
2. In the recent years, we have been moving to more narrative driven examples, where text to place together with code. For example: [Common pitfalls in the interpretation of coefficients of linear models](https://scikit-learn.org/stable/auto_examples/inspection/plot_linear_model_coefficient_interpretation.html#sphx-glr-auto-examples-inspection-plot-linear-model-coefficient-interpretation-py). This means moving text from the big paragraph in the beginning and placing them into the text to create a narrative.

sergeyf commented:
Thanks! I can make those changes. What regression dataset do you recommend to replace Boston? Smaller is best because this example is hefty, but I can always subsample any dataset.

nxorable commented:
fetch_california is similar to load_boston, albeit larger

On Mon, Jan 3, 2022 at 10:12 PM Sergey Feldman ***@***.***>
wrote:

> Thanks! I can make those changes. What regression dataset do you recommend
> to replace Boston? Smaller is best because this example is hefty, but I can
> always subsample any dataset.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/13025#issuecomment-1004496569>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABURKX3GKTIT3F6WM3WK5DDUUJQQNANCNFSM4GRNRPRA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

glemaitre commented:
> Any idea if we can get this merged once all the tests pass?

I would like to see how this example integrates within the proposal in https://github.com/scikit-learn/scikit-learn/issues/21967.

sergeyf commented:
@glemaitre MICE is in the family of multiple imputation - perform imputation multiple times, then apply your subsequent pipeline multiple times also, and then have multiple solutions. For `sklearn` users the subsequent pipeline will often be "train/val/test a ML alg". I read through #21967 and multiple imputation isn't mentioned at all, but it is common in the stats world as far as I understand. This example would be useful because people coming from stats might want to do what the `mice` R package does, but don't know how with `IterativeImputer` as it does not do it out of the box. 

To summarize:
- #21967 is about how to make a guide to imputation, (mostly) for the purpose of ML.
- This example is about multiple imputation.
- The two don't have an clear bridge at the moment.

glemaitre commented:
I agree that having an example defining what is "multiple imputations" is important to remove the confusion with the iterative procedure of the `IterativeImputer` and that there is no confusion. In this regard, I find the example too complex.

In this regard, I would prefer to have a single pipeline to make single imputation and then create a specific estimator to show how to make multiple imputations. We would not even need to use an `IterativeImputer` in this case. This would speed up the example that is currently taking up to 3 minutes while we usually try to have examples running under 30 seconds.

I think that this is super important to point out in the discussion that the example stands at providing a definition of "multiple imputations with code" rather than showing that multiple imputations work better. I am not sure that currently in ML setting there is any evidence that multiple imputations are working better than using a strong learner (@GaelVaroquaux and @marineLM have better insights than me on this).

GaelVaroquaux commented:
Cc @a-pl (we need to put your paper on HAL)

sergeyf commented:
I'm a bit confused. We do have a single pipeline in example 2: https://github.com/sergeyf/scikit-learn/blob/iterativeimputer_mice_example/examples/impute/plot_multiple_imputation.py#L303

And it's used multiple times to do MICE: https://github.com/sergeyf/scikit-learn/blob/iterativeimputer_mice_example/examples/impute/plot_multiple_imputation.py#L315

Can you please clarify what you'd like changed?
